{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center>\n",
    "    <img src=\"https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png\"  width=\"300\" alt=\"cognitiveclass.ai logo\"  />\n",
    "</center>\n",
    "\n",
    "<h1 align=center><font size = 5>Assignment: SQL Notebook for Peer Assignment</font></h1>\n",
    "\n",
    "Estimated time needed: **60** minutes.\n",
    "\n",
    "## Introduction\n",
    "\n",
    "Using this Python notebook you will:\n",
    "\n",
    "1.  Understand the Spacex DataSet\n",
    "2.  Load the dataset  into the corresponding table in a Db2 database\n",
    "3.  Execute SQL queries to answer assignment questions\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview of the DataSet\n",
    "\n",
    "SpaceX has gained worldwide attention for a series of historic milestones.\n",
    "\n",
    "It is the only private company ever to return a spacecraft from low-earth orbit, which it first accomplished in December 2010.\n",
    "SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars wheras other providers cost upward of 165 million dollars each, much of the savings is because Space X can reuse the first stage.\n",
    "\n",
    "Therefore if we can determine if the first stage will land, we can determine the cost of a launch.\n",
    "\n",
    "This information can be used if an alternate company wants to bid against SpaceX for a rocket launch.\n",
    "\n",
    "This dataset includes a record for each payload carried during a SpaceX mission into outer space.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download the datasets\n",
    "\n",
    "This assignment requires you to load the spacex dataset.\n",
    "\n",
    "In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. Click on the link below to download and save the dataset (.CSV file):\n",
    "\n",
    "<a href=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01\" target=\"_blank\">Spacex DataSet</a>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Store the dataset in database table\n",
    "\n",
    "**it is highly recommended to manually load the table using the database console LOAD tool in DB2**.\n",
    "\n",
    "<img src = \"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/images/spacexload.png\">\n",
    "\n",
    "Now open the Db2 console, open the LOAD tool, Select / Drag the .CSV file for the  dataset, Next create a New Table, and then follow the steps on-screen instructions to load the data. Name the new table as follows:\n",
    "\n",
    "**SPACEXDATASET**\n",
    "\n",
    "**Follow these steps while using old DB2 UI which is having Open Console Screen**\n",
    "\n",
    "**Note:While loading Spacex dataset, ensure that detect datatypes is disabled. Later click on the pencil icon(edit option).**\n",
    "\n",
    "1.  Change the Date Format by manually typing DD-MM-YYYY and timestamp format as DD-MM-YYYY HH\\:MM:SS\n",
    "\n",
    "2.  Change the PAYLOAD_MASS\\_\\_KG\\_  datatype  to INTEGER.\n",
    "\n",
    "<img src = \"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/images/spacexload2.png\">\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Changes to be considered when having DB2 instance with the new UI having Go to UI screen**\n",
    "\n",
    "*   Refer to this insruction in this <a href=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Sign%20up%20for%20IBM%20Cloud%20-%20Create%20Db2%20service%20instance%20-%20Get%20started%20with%20the%20Db2%20console/instructional-labs.md.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01\">link</a> for viewing  the new  Go to UI screen.\n",
    "\n",
    "*   Later click on **Data link(below SQL)**  in the Go to UI screen  and click on **Load Data** tab.\n",
    "\n",
    "*   Later browse for the downloaded spacex file.\n",
    "\n",
    "<img src=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/images/browsefile.png\" width=\"800\"/>\n",
    "\n",
    "*   Once done select the schema andload the file.\n",
    "\n",
    " <img src=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/images/spacexload3.png\" width=\"800\"/>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: sqlalchemy in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (2.0.41)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from sqlalchemy) (4.12.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: ibm_db_sa in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (0.4.1)\n",
      "Requirement already satisfied: sqlalchemy>=0.7.3 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ibm_db_sa) (2.0.41)\n",
      "Requirement already satisfied: ibm-db>=2.0.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ibm_db_sa) (3.2.6)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from sqlalchemy>=0.7.3->ibm_db_sa) (4.12.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: ipython-sql in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (0.5.0)\n",
      "Requirement already satisfied: prettytable in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython-sql) (3.14.0)\n",
      "Requirement already satisfied: ipython in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython-sql) (8.18.1)\n",
      "Requirement already satisfied: sqlalchemy>=2.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython-sql) (2.0.41)\n",
      "Requirement already satisfied: sqlparse in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython-sql) (0.5.3)\n",
      "Requirement already satisfied: six in /Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/site-packages (from ipython-sql) (1.15.0)\n",
      "Requirement already satisfied: ipython-genutils in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython-sql) (0.2.0)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from sqlalchemy>=2.0->ipython-sql) (4.12.2)\n",
      "Requirement already satisfied: decorator in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (0.19.2)\n",
      "Requirement already satisfied: matplotlib-inline in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (0.1.7)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (3.0.50)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (2.19.1)\n",
      "Requirement already satisfied: stack-data in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (0.6.3)\n",
      "Requirement already satisfied: traitlets>=5 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (5.14.3)\n",
      "Requirement already satisfied: exceptiongroup in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (1.2.2)\n",
      "Requirement already satisfied: pexpect>4.3 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from ipython->ipython-sql) (4.9.0)\n",
      "Requirement already satisfied: wcwidth in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from prettytable->ipython-sql) (0.2.13)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from jedi>=0.16->ipython->ipython-sql) (0.8.4)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from pexpect>4.3->ipython->ipython-sql) (0.7.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from stack-data->ipython->ipython-sql) (2.2.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from stack-data->ipython->ipython-sql) (3.0.0)\n",
      "Requirement already satisfied: pure-eval in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (from stack-data->ipython->ipython-sql) (0.2.3)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pandas in /Users/josebautearellan/Library/Python/3.9/lib/python/site-packages (2.2.3)\n",
      "\u001b[31mERROR: Could not find a version that satisfies the requirement as (from versions: none)\u001b[0m\u001b[31m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "\u001b[31mERROR: No matching distribution found for as\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!pip install sqlalchemy\n",
    "!pip install ibm_db_sa\n",
    "!pip install ipython-sql\n",
    "!pip install pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Connect to the database\n",
    "\n",
    "Let us first load the SQL extension and establish a connection with the database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    }
   ],
   "source": [
    "%load_ext sql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**DB2 magic in case of old UI service credentials.**\n",
    "\n",
    "In the next cell enter your db2 connection string. Recall you created Service Credentials for your Db2 instance before. From the **uri** field of your Db2 service credentials copy everything after db2:// (except the double quote at the end) and paste it in the cell below after ibm_db_sa://\n",
    "\n",
    "<img src =\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_edX/images/URI.jpg\">\n",
    "\n",
    "in the following format\n",
    "\n",
    "**%sql ibm_db_sa://my-username:my-password\\@my-hostname:my-port/my-db-name**\n",
    "\n",
    "**DB2 magic in case of new UI service credentials.**\n",
    "\n",
    "<img src =\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/images/servicecredentials.png\" width=600>  \n",
    "\n",
    "*   Use the following format.\n",
    "\n",
    "*   Add security=SSL at the end\n",
    "\n",
    "**%sql ibm_db_sa://my-username:my-password\\@my-hostname:my-port/my-db-name?security=SSL**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/ibm_db_dbi.py\", line 736, in connect\n",
      "    conn = ibm_db.connect(dsn, '', '', conn_options)\n",
      "Exception: [IBM][CLI Driver] SQL30082N  Security processing failed with reason \"24\" (\"USERNAME AND/OR PASSWORD INVALID\").  SQLSTATE=08001 SQLCODE=-30082\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 145, in __init__\n",
      "    self._dbapi_connection = engine.raw_connection()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 3297, in raw_connection\n",
      "    return self.pool.connect()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 449, in connect\n",
      "    return _ConnectionFairy._checkout(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 1264, in _checkout\n",
      "    fairy = _ConnectionRecord.checkout(pool)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 713, in checkout\n",
      "    rec = pool._do_get()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/impl.py\", line 180, in _do_get\n",
      "    self._dec_overflow()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/util/langhelpers.py\", line 224, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/impl.py\", line 177, in _do_get\n",
      "    return self._create_connection()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 390, in _create_connection\n",
      "    return _ConnectionRecord(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 675, in __init__\n",
      "    self.__connect()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 902, in __connect\n",
      "    pool.logger.debug(\"Error on connect(): %s\", e)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/util/langhelpers.py\", line 224, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 897, in __connect\n",
      "    self.dbapi_connection = connection = pool._invoke_creator(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/create.py\", line 646, in connect\n",
      "    return dialect.connect(*cargs, **cparams)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/default.py\", line 625, in connect\n",
      "    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/ibm_db_dbi.py\", line 744, in connect\n",
      "    raise _get_exception(inst)\n",
      "ibm_db_dbi.OperationalError: ibm_db_dbi::OperationalError: [IBM][CLI Driver] SQL30082N  Security processing failed with reason \"24\" (\"USERNAME AND/OR PASSWORD INVALID\").  SQLSTATE=08001 SQLCODE=-30082\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sql/magic.py\", line 196, in execute\n",
      "    conn = sql.connection.Connection.set(\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sql/connection.py\", line 70, in set\n",
      "    cls.current = existing or Connection(descriptor, connect_args, creator)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sql/connection.py\", line 55, in __init__\n",
      "    self.internal_connection = engine.connect()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 3273, in connect\n",
      "    return self._connection_cls(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 147, in __init__\n",
      "    Connection._handle_dbapi_exception_noconnection(\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 2436, in _handle_dbapi_exception_noconnection\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 145, in __init__\n",
      "    self._dbapi_connection = engine.raw_connection()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/base.py\", line 3297, in raw_connection\n",
      "    return self.pool.connect()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 449, in connect\n",
      "    return _ConnectionFairy._checkout(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 1264, in _checkout\n",
      "    fairy = _ConnectionRecord.checkout(pool)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 713, in checkout\n",
      "    rec = pool._do_get()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/impl.py\", line 180, in _do_get\n",
      "    self._dec_overflow()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/util/langhelpers.py\", line 224, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/impl.py\", line 177, in _do_get\n",
      "    return self._create_connection()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 390, in _create_connection\n",
      "    return _ConnectionRecord(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 675, in __init__\n",
      "    self.__connect()\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 902, in __connect\n",
      "    pool.logger.debug(\"Error on connect(): %s\", e)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/util/langhelpers.py\", line 224, in __exit__\n",
      "    raise exc_value.with_traceback(exc_tb)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/pool/base.py\", line 897, in __connect\n",
      "    self.dbapi_connection = connection = pool._invoke_creator(self)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/create.py\", line 646, in connect\n",
      "    return dialect.connect(*cargs, **cparams)\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/sqlalchemy/engine/default.py\", line 625, in connect\n",
      "    return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501\n",
      "  File \"/Users/josebautearellan/Library/Python/3.9/lib/python/site-packages/ibm_db_dbi.py\", line 744, in connect\n",
      "    raise _get_exception(inst)\n",
      "sqlalchemy.exc.OperationalError: (ibm_db_dbi.OperationalError) ibm_db_dbi::OperationalError: [IBM][CLI Driver] SQL30082N  Security processing failed with reason \"24\" (\"USERNAME AND/OR PASSWORD INVALID\").  SQLSTATE=08001 SQLCODE=-30082\n",
      "(Background on this error at: https://sqlalche.me/e/20/e3q8)\n",
      "\n",
      "Connection info needed in SQLAlchemy format, example:\n",
      "               postgresql://username:password@hostname/dbname\n",
      "               or an existing connection: dict_keys([])\n"
     ]
    }
   ],
   "source": [
    "%sql ibm_db_sa://wzf08322:ueXYepcYNnElQRV6@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/bludb?security=SSL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "# Leer el archivo CSV en un DataFrame de pandas\n",
    "df_spacex = pd.read_csv(ruta_archivo_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tasks\n",
    "\n",
    "Now write and execute SQL queries to solve the assignment tasks.\n",
    "\n",
    "### Task 1\n",
    "\n",
    "##### Display the names of the unique launch sites  in the space mission\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Launch_Site\n",
      " CCAFS LC-40\n",
      " VAFB SLC-4E\n",
      "  KSC LC-39A\n",
      "CCAFS SLC-40\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Obtener los valores únicos de la columna 'Launch_Site'\n",
    "    unique_launch_sites_df = pd.DataFrame({'Launch_Site': df_spacex['Launch_Site'].unique()})\n",
    "\n",
    "    # Mostrar el DataFrame en formato de tabla\n",
    "    print(unique_launch_sites_df.to_string(index=False))\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2\n",
    "\n",
    "##### Display 5 records where launch sites begin with the string 'CCA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         Date Time (UTC) Booster_Version  Launch_Site                                                        Payload  PAYLOAD_MASS__KG_      Orbit         Customer Mission_Outcome      Landing_Outcome\n",
      "0  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40                           Dragon Spacecraft Qualification Unit                  0        LEO           SpaceX         Success  Failure (parachute)\n",
      "1  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40  Dragon demo flight C1, two CubeSats, barrel of Brouere cheese                  0  LEO (ISS)  NASA (COTS) NRO         Success  Failure (parachute)\n",
      "2  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40                                          Dragon demo flight C2                525  LEO (ISS)      NASA (COTS)         Success           No attempt\n",
      "3  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40                                                   SpaceX CRS-1                500  LEO (ISS)       NASA (CRS)         Success           No attempt\n",
      "4  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40                                                   SpaceX CRS-2                677  LEO (ISS)       NASA (CRS)         Success           No attempt\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Filtrar las filas donde la columna 'Launch_Site' comienza con 'CCA'\n",
    "    cca_launch_sites = df_spacex[df_spacex['Launch_Site'].str.startswith('CCA')]\n",
    "\n",
    "    # Mostrar los primeros 5 registros del DataFrame filtrado\n",
    "    print(cca_launch_sites.head(5).to_string())\n",
    "    \n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3\n",
    "\n",
    "##### Display the total payload mass carried by boosters launched by NASA (CRS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total payload mass carried by boosters launched by NASA (CRS): 45596 KG\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Filtrar las misiones where the 'Customer' is 'NASA (CRS)'\n",
    "    nasa_crs_missions = df_spacex[df_spacex['Customer'] == 'NASA (CRS)']\n",
    "\n",
    "    # Calculate the total payload mass for these missions\n",
    "    total_payload_nasa_crs = nasa_crs_missions['PAYLOAD_MASS__KG_'].sum()\n",
    "\n",
    "    # Display the result\n",
    "    print(f\"Total payload mass carried by boosters launched by NASA (CRS): {total_payload_nasa_crs} KG\")\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que el nombre de la columna es correcto.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 4\n",
    "\n",
    "##### Display average payload mass carried by booster version F9 v1.1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average payload mass carried by booster version F9 v1.1: 2928.40 KG\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Filter for missions with booster version 'F9 v1.1'\n",
    "    f9_v1_1_boosters = df_spacex[df_spacex['Booster_Version'] == 'F9 v1.1']\n",
    "\n",
    "    # Calculate the average payload mass for these missions\n",
    "    average_payload_f9_v1_1 = f9_v1_1_boosters['PAYLOAD_MASS__KG_'].mean()\n",
    "\n",
    "    # Display the result\n",
    "    if not pd.isna(average_payload_f9_v1_1):\n",
    "        print(f\"Average payload mass carried by booster version F9 v1.1: {average_payload_f9_v1_1:.2f} KG\")\n",
    "    else:\n",
    "        print(\"No launches found with booster version F9 v1.1 to calculate the average payload mass.\")\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que los nombres de las columnas 'Booster_Version' y 'PAYLOAD_MASS__KG_' son correctos.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 5\n",
    "\n",
    "##### List the date when the first succesful landing outcome in ground pad was acheived.\n",
    "\n",
    "*Hint:Use min function*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The date of the first successful landing outcome in ground pad was: 2015-12-22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/d4/0j6037cj5t78ml2jnxrpv3mr0000gn/T/ipykernel_62908/3682545468.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  successful_ground_pad['Date'] = pd.to_datetime(successful_ground_pad['Date'])\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Filter for successful landing outcomes in ground pad\n",
    "    successful_ground_pad = df_spacex[df_spacex['Landing_Outcome'] == 'Success (ground pad)']\n",
    "\n",
    "    # Sort the DataFrame by date to find the first successful landing\n",
    "    successful_ground_pad['Date'] = pd.to_datetime(successful_ground_pad['Date'])\n",
    "    successful_ground_pad_sorted = successful_ground_pad.sort_values(by='Date')\n",
    "\n",
    "    # Get the date of the first successful landing\n",
    "    if not successful_ground_pad_sorted.empty:\n",
    "        first_successful_landing_date = successful_ground_pad_sorted.iloc[0]['Date'].strftime('%Y-%m-%d')\n",
    "        print(f\"The date of the first successful landing outcome in ground pad was: {first_successful_landing_date}\")\n",
    "    else:\n",
    "        print(\"No successful landing outcomes in ground pad were found in the data.\")\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que las columnas 'Landing_Outcome' and 'Date' existen.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * ibm_db_sa://wzf08322:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/bludb\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>first_successful_landing</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2015-12-22</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(datetime.date(2015, 12, 22),)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select min(date) as first_successful_landing from SPACEXDATASET where landing__outcome = 'Success (ground pad)';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 6\n",
    "\n",
    "##### List the names of the boosters which have success in drone ship and have payload mass greater than 4000 but less than 6000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Boosters with successful drone ship landing and payload mass between 4000 and 6000 KG:\n",
      "F9 FT B1022\n",
      "F9 FT B1026\n",
      "F9 FT  B1021.2\n",
      "F9 FT  B1031.2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Filter for successful landing on a drone ship\n",
    "    successful_drone_ship = df_spacex[df_spacex['Landing_Outcome'] == 'Success (drone ship)']\n",
    "\n",
    "    # Filter for payload mass greater than 4000 and less than 6000 KG\n",
    "    payload_filtered = successful_drone_ship[(successful_drone_ship['PAYLOAD_MASS__KG_'] > 4000) & (successful_drone_ship['PAYLOAD_MASS__KG_'] < 6000)]\n",
    "\n",
    "    # List the names of the boosters (assuming 'Booster_Version' contains the booster name)\n",
    "    booster_names = payload_filtered['Booster_Version'].tolist()\n",
    "\n",
    "    if booster_names:\n",
    "        print(\"Boosters with successful drone ship landing and payload mass between 4000 and 6000 KG:\")\n",
    "        for booster in booster_names:\n",
    "            print(booster)\n",
    "    else:\n",
    "        print(\"No boosters found with successful drone ship landing and payload mass between 4000 and 6000 KG.\")\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que las columnas 'Landing_Outcome', 'PAYLOAD_MASS__KG_', and 'Booster_Version' existen.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * ibm_db_sa://wzf08322:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/bludb\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>booster_version</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>F9 FT B1022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>F9 FT B1026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>F9 FT  B1021.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>F9 FT  B1031.2</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('F9 FT B1022',), ('F9 FT B1026',), ('F9 FT  B1021.2',), ('F9 FT  B1031.2',)]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select booster_version from SPACEXDATASET where landing__outcome = 'Success (drone ship)' and payload_mass__kg_ between 4000 and 6000;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 7\n",
    "\n",
    "##### List the total number of successful and failure mission outcomes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 Mission_Outcome  total_number\n",
      "             Failure (in flight)             1\n",
      "                         Success            99\n",
      "Success (payload status unclear)             1\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Clean the 'Mission_Outcome' column by removing leading/trailing whitespace\n",
    "    df_spacex['Mission_Outcome'] = df_spacex['Mission_Outcome'].str.strip()\n",
    "\n",
    "    # Group by 'Mission_Outcome' and count the occurrences\n",
    "    mission_outcome_counts = df_spacex.groupby('Mission_Outcome').size().reset_index(name='total_number')\n",
    "\n",
    "    # Display the result as a table\n",
    "    print(mission_outcome_counts.to_string(index=False))\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que la columna 'Mission_Outcome' existe.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * ibm_db_sa://wzf08322:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/bludb\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>mission_outcome</th>\n",
       "        <th>total_number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Failure (in flight)</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Success</td>\n",
       "        <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Success (payload status unclear)</td>\n",
       "        <td>1</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('Failure (in flight)', 1),\n",
       " ('Success', 99),\n",
       " ('Success (payload status unclear)', 1)]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select mission_outcome, count(*) as total_number from SPACEXDATASET group by mission_outcome;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 8\n",
    "\n",
    "##### List the   names of the booster_versions which have carried the maximum payload mass. Use a subquery\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Booster versions which carried the maximum payload mass:\n",
      "F9 B5 B1048.4\n",
      "F9 B5 B1049.4\n",
      "F9 B5 B1051.3\n",
      "F9 B5 B1056.4\n",
      "F9 B5 B1048.5\n",
      "F9 B5 B1051.4\n",
      "F9 B5 B1049.5\n",
      "F9 B5 B1060.2 \n",
      "F9 B5 B1058.3 \n",
      "F9 B5 B1051.6\n",
      "F9 B5 B1060.3\n",
      "F9 B5 B1049.7 \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Find the maximum payload mass using a subquery-like approach\n",
    "    max_payload = df_spacex['PAYLOAD_MASS__KG_'].max()\n",
    "\n",
    "    # Filter the DataFrame to get booster versions with the maximum payload mass\n",
    "    boosters_with_max_payload = df_spacex[df_spacex['PAYLOAD_MASS__KG_'] == max_payload]['Booster_Version'].unique().tolist()\n",
    "\n",
    "    # Display the names of the booster versions\n",
    "    if boosters_with_max_payload:\n",
    "        print(\"Booster versions which carried the maximum payload mass:\")\n",
    "        for booster in boosters_with_max_payload:\n",
    "            print(booster)\n",
    "    else:\n",
    "        print(\"No booster versions found with payload mass data.\")\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que las columnas 'Booster_Version' y 'PAYLOAD_MASS__KG_' existen.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 9\n",
    "\n",
    "##### List the failed landing_outcomes in drone ship, their booster versions, and launch site names for the in year 2015\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Month       Date Booster_Version Launch_Site      Landing_Outcome\n",
      "January 2015-01-10   F9 v1.1 B1012 CCAFS LC-40 Failure (drone ship)\n",
      "  April 2015-04-14   F9 v1.1 B1015 CCAFS LC-40 Failure (drone ship)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Convert the 'Date' column to datetime objects\n",
    "    df_spacex['Date'] = pd.to_datetime(df_spacex['Date'])\n",
    "\n",
    "    # Filter for failed drone ship landings in the year 2015\n",
    "    failed_drone_ship_2015 = df_spacex[\n",
    "        (df_spacex['Landing_Outcome'] == 'Failure (drone ship)') & (df_spacex['Date'].dt.year == 2015)\n",
    "    ]\n",
    "\n",
    "    # Extract the required columns\n",
    "    result = failed_drone_ship_2015[['Date', 'Booster_Version', 'Launch_Site', 'Landing_Outcome']].copy()\n",
    "\n",
    "    # Format the 'Date' column to display the month name\n",
    "    result['Month'] = result['Date'].dt.month_name()\n",
    "\n",
    "    # Reorder the columns to match the SQL query output\n",
    "    result = result[['Month', 'Date', 'Booster_Version', 'Launch_Site', 'Landing_Outcome']]\n",
    "\n",
    "    # Display the result\n",
    "    print(result.to_string(index=False))\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que las columnas 'Date', 'Booster_Version', 'Launch_Site', y 'Landing_Outcome' existen.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * ibm_db_sa://wzf08322:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/bludb\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>MONTH</th>\n",
       "        <th>DATE</th>\n",
       "        <th>booster_version</th>\n",
       "        <th>launch_site</th>\n",
       "        <th>landing__outcome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>January</td>\n",
       "        <td>2015-01-10</td>\n",
       "        <td>F9 v1.1 B1012</td>\n",
       "        <td>CCAFS LC-40</td>\n",
       "        <td>Failure (drone ship)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>April</td>\n",
       "        <td>2015-04-14</td>\n",
       "        <td>F9 v1.1 B1015</td>\n",
       "        <td>CCAFS LC-40</td>\n",
       "        <td>Failure (drone ship)</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('January', datetime.date(2015, 1, 10), 'F9 v1.1 B1012', 'CCAFS LC-40', 'Failure (drone ship)'),\n",
       " ('April', datetime.date(2015, 4, 14), 'F9 v1.1 B1015', 'CCAFS LC-40', 'Failure (drone ship)')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql select monthname(date) as month, date, booster_version, launch_site, landing__outcome from SPACEXDATASET\n",
    "      where landing__outcome = 'Failure (drone ship)' and year(date)=2015;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 10\n",
    "\n",
    "##### Rank the count of landing outcomes (such as Failure (drone ship) or Success (ground pad)) between the date 2010-06-04 and 2017-03-20, in descending order\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Landing_Outcome  count_outcomes\n",
      "            No attempt              10\n",
      "  Failure (drone ship)               5\n",
      "  Success (drone ship)               5\n",
      "    Controlled (ocean)               3\n",
      "  Success (ground pad)               3\n",
      "   Failure (parachute)               2\n",
      "  Uncontrolled (ocean)               2\n",
      "Precluded (drone ship)               1\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ruta_archivo_csv = '/Users/josebautearellan/Desktop/IBM/Data scientist /Databases and SQL for Data Science with Python/Proyecto/Spacex.csv'\n",
    "\n",
    "try:\n",
    "    # Leer el archivo CSV en un DataFrame de pandas con UTF-8 encoding\n",
    "    df_spacex = pd.read_csv(ruta_archivo_csv, encoding='utf-8')\n",
    "\n",
    "    # Convert the 'Date' column to datetime objects\n",
    "    df_spacex['Date'] = pd.to_datetime(df_spacex['Date'])\n",
    "\n",
    "    # Define the date range\n",
    "    start_date = '2010-06-04'\n",
    "    end_date = '2017-03-20'\n",
    "\n",
    "    # Filter data within the specified date range\n",
    "    date_filtered = df_spacex[(df_spacex['Date'] >= start_date) & (df_spacex['Date'] <= end_date)]\n",
    "\n",
    "    # Group by 'Landing_Outcome' and count the occurrences\n",
    "    landing_outcome_counts = date_filtered.groupby('Landing_Outcome').size().reset_index(name='count_outcomes')\n",
    "\n",
    "    # Sort the results by 'count_outcomes' in descending order\n",
    "    sorted_outcomes = landing_outcome_counts.sort_values(by='count_outcomes', ascending=False)\n",
    "\n",
    "    # Display the result as a table\n",
    "    print(sorted_outcomes.to_string(index=False))\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: El archivo no se encontró en la ruta: {ruta_archivo_csv}\")\n",
    "except KeyError as e:\n",
    "    print(f\"Error: La columna '{e}' no se encuentra en el archivo CSV. Asegúrate de que las columnas 'Date' y 'Landing_Outcome' existen.\")\n",
    "except Exception as e:\n",
    "    print(f\"Ocurrió un error al leer o procesar el archivo CSV: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference Links\n",
    "\n",
    "*   <a href =\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20String%20Patterns%20-%20Sorting%20-%20Grouping/instructional-labs.md.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01&origin=www.coursera.org\">Hands-on Lab : String Patterns, Sorting and Grouping</a>\n",
    "\n",
    "*   <a  href=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Built-in%20functions%20/Hands-on_Lab__Built-in_Functions.md.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01&origin=www.coursera.org\">Hands-on Lab: Built-in functions</a>\n",
    "\n",
    "*   <a  href=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Sub-queries%20and%20Nested%20SELECTs%20/instructional-labs.md.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01&origin=www.coursera.org\">Hands-on Lab : Sub-queries and Nested SELECT Statements</a>\n",
    "\n",
    "*   <a href=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-3-SQLmagic.ipynb?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01\">Hands-on Tutorial: Accessing Databases with SQL magic</a>\n",
    "\n",
    "*   <a href= \"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-4-Analyzing.ipynb?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01\">Hands-on Lab: Analyzing a real World Data Set</a>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Author(s)\n",
    "\n",
    "<h4> Lakshmi Holla </h4>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other Contributors\n",
    "\n",
    "<h4> Rav Ahuja </h4>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change log\n",
    "\n",
    "| Date       | Version | Changed by    | Change Description        |\n",
    "| ---------- | ------- | ------------- | ------------------------- |\n",
    "| 2021-07-09 | 0.2     | Lakshmi Holla | Changes made in magic sql |\n",
    "| 2021-05-20 | 0.1     | Lakshmi Holla | Created Initial Version   |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <h3 align=\"center\"> © IBM Corporation 2021. All rights reserved. <h3/>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
